src/caches/abstractbeliefpropagationcache.jl

```diff
@@ -303,24 +338,18 @@ end
 """
 More generic interface for update, with default params
 """
-function update(
-  alg::Algorithm,
-  bpc::AbstractBeliefPropagationCache;
-  edges=default_edge_sequence(alg, bpc),
-  maxiter=default_bp_maxiter(alg, bpc),
-  message_update_kwargs=default_message_update_kwargs(alg, bpc),
-  tol=nothing,
-  verbose=false,
-)
-  compute_error = !isnothing(tol)
-  if isnothing(maxiter)
+function update(alg::Algorithm, bpc::AbstractBeliefPropagationCache; kwargs...)
+  compute_error = !isnothing(alg.kwargs.tol)
+  if isnothing(alg.kwargs.maxiter)
     error("You need to specify a number of iterations for BP!")
   end
-  for i in 1:maxiter
+  for i in 1:alg.kwargs.maxiter
     diff = compute_error ? Ref(0.0) : nothing
-    bpc = update(alg, bpc, edges; (update_diff!)=diff, message_update_kwargs...)
-    if compute_error && (diff.x / length(edges)) <= tol
-      if verbose
+    bpc = update_one_iteration(
+      alg, bpc, alg.kwargs.edge_sequence; (update_diff!)=diff, kwargs...
```

Comment:
I see that `update_diff!` is still being passed along, that's slightly unfortunate since it forces every implementation of `update_one_iteration`/`update_iteration` to compute that. Can't we compute it directly in this function by comparing `bpc` before and after calling `update_iteration`?

Reply:
Yeah I think that is the only one being passed around anymore. It could be computed inside the function, although I think that forces us to have two copies of all of the message tensors (one for this iteration and one for the one before) which is more memory intensive.

I think currently we compute update_diff `message` by `message` as they are updated, which means we should only ever have the current messages and one of the previous messages if the garbage collection is done right

Reply:
Probably in the cases where keeping two copies of the BP cache in memory is too much, it would be better to write parts of it to disk anyway. Relying on the garbage collector could be finicky.

Reply:
Yeah that's a good point. I have left it as is for now and made sure the `kwarg` `update_diff` is specified explicitly  and no `kwargs` are passed further than `update_iteration` but in the future going to disk sounds sensible